src/dtls13.c

```diff
@@ -0,0 +1,996 @@
+                                            word32 frag_offset,
+                                            word32 frag_length,
+                                            word32 msg_length)
+{
+    struct dtls13_handshake_header *hdr;
+
+    hdr = (struct dtls13_handshake_header *)output;
+
+    hdr->msg_type = msg_type;
+    c32to24((word32)msg_length, hdr->length);
+    c16toa(ssl->keys.dtls_handshake_number, hdr->message_seq);
+
+    c32to24(frag_offset, hdr->fragment_offset);
+    c32to24(frag_length, hdr->fragment_length);
+
+    return 0;
+}
+
+static int dtls13_type_is_encrypted(enum HandShakeType hs_type)
+{
+    switch (hs_type) {
+    case hello_request:
+    case hello_verify_request:
+    case client_hello:
+    case server_hello:
+        return 0;
+    case encrypted_extensions:
+    case session_ticket:
+    case end_of_early_data   :
+    case hello_retry_request :
+    case certificate         :
+    case server_key_exchange :
+    case certificate_request :
+    case server_hello_done   :
+    case certificate_verify  :
+    case client_key_exchange :
+    case finished            :
+    case certificate_status  :
+    case key_update          :
+    case change_cipher_hs    :
+    case message_hash        :
+    case no_shake            :
+        return 1;
+    }
+
+    /* abide compilers */
+    return 0;
+}
+
+
+static int dtls13_get_rn_mask(WOLFSSL *ssl, byte *ciphertext, byte *mask,
+                              word16 side)
+{
+    byte *key;
+    Aes aes;
+    int ret;
+
+    /* FIXME: implement other ciphers */
+    if (ssl->specs.bulk_cipher_algorithm != wolfssl_aes_gcm)
+        return WOLFSSL_NOT_IMPLEMENTED;
+
+    ret = wc_AesInit(&aes, ssl->heap, -1);
+    if (ret != 0)
+        return ret;
+
+    key = (side == WOLFSSL_SERVER_END) ?
+        ssl->keys.server_sn_key : ssl->keys.client_sn_key;
+
+    ret = wc_AesSetKey(&aes, key, ssl->specs.key_size, NULL, AES_ENCRYPTION);
+    if (ret != 0)
+        goto exit;
+
+    ret = wc_AesEncryptDirect(&aes, mask, ciphertext);
+    if (ret != 0)
+        goto exit;
+
+ exit:
+    wc_AesFree(&aes);
+    return ret;
+}
+
+static int dtls13_encrypt_decrypt_record_number(WOLFSSL *ssl,
+                                                byte *seq,
+                                                int seq_length,
+                                                byte *ciphertext,
+                                                enum rn_direction dir)
+
+{
+    byte mask[RN_MASK_SIZE];
+    int ret, i;
+    word8 side;
+
+    if (dir == PROTECT)
+        side = ssl->options.side;
+    else
+        side = (ssl->options.side == WOLFSSL_SERVER_END) ?
+            WOLFSSL_CLIENT_END : WOLFSSL_SERVER_END;
+
+    ret = dtls13_get_rn_mask(ssl, ciphertext, mask, side);
+    if (ret != 0)
+        return ret;
+
+    for (i = 0; i < seq_length; ++i)
+        seq[i] ^= mask[i];
+
+    return 0;
+}
+
+/* this is a clone of DtlsMsgDrain */
```

Comment:
Any reason they can't be consolidated into one implementation?

Reply:
Not at the moment, I'll merge them if I can